### ISSUE TITLE
No way to list common rails commands from CLI outside a rails app directory

### ISSUE THREAD

**Issue by olivierlacan** [rails/rails#40823]
### Steps to reproduce
- `rails --help`
- `rails help`
- `rails -h`

### Expected behavior

Using `rails --help` should print usage instructions for common Rails commands.

```
rails --help
```

<details>
<summary>Command output</summary>
<pre>
The most common rails commands are:
 generate     Generate new code (short-cut alias: "g")
 console      Start the Rails console (short-cut alias: "c")
 server       Start the Rails server (short-cut alias: "s")
 test         Run tests except system tests (short-cut alias: "t")
 test:system  Run system tests
 dbconsole    Start a console for the database specified in config/database.yml
              (short-cut alias: "db")

All commands can be run with -h (or --help) for more information.
In addition to those commands, there are:
</pre>
</details>

### Actual behavior

It prints usage instructions specific to `rails new` instead.

```
rails --help
```

<details>
<summary>Command output</summary>
<pre>
Usage:
  rails new APP_PATH [options]

Options:
      [--skip-namespace], [--no-skip-namespace]              # Skip namespace (affects only isolated engines)
      [--skip-collision-check], [--no-skip-collision-check]  # Skip collision check
  -r, [--ruby=PATH]                                          # Path to the Ruby binary of your choice
                                                             # Default: /Users/olivierlacan/.rbenv/versions/2.7.2/bin/ruby
  -m, [--template=TEMPLATE]                                  # Path to some application template (can be a filesystem path or URL)
  -d, [--database=DATABASE]                                  # Preconfigure for selected database (options: mysql/postgresql/sqlite3/oracle/sqlserver/jdbcmysql/jdbcsqlite3/jdbcpostgresql/jdbc)
                                                             # Default: sqlite3
      [--skip-gemfile], [--no-skip-gemfile]                  # Don't create a Gemfile
  -G, [--skip-git], [--no-skip-git]                          # Skip .gitignore file
      [--skip-keeps], [--no-skip-keeps]                      # Skip source control .keep files
  -M, [--skip-action-mailer], [--no-skip-action-mailer]      # Skip Action Mailer files
      [--skip-action-mailbox], [--no-skip-action-mailbox]    # Skip Action Mailbox gem
      [--skip-action-text], [--no-skip-action-text]          # Skip Action Text gem
  -O, [--skip-active-record], [--no-skip-active-record]      # Skip Active Record files
      [--skip-active-job], [--no-skip-active-job]            # Skip Active Job
      [--skip-active-storage], [--no-skip-active-storage]    # Skip Active Storage files
  -P, [--skip-puma], [--no-skip-puma]                        # Skip Puma related files
  -C, [--skip-action-cable], [--no-skip-action-cable]        # Skip Action Cable files
  -S, [--skip-sprockets], [--no-skip-sprockets]              # Skip Sprockets files
      [--skip-spring], [--no-skip-spring]                    # Don't install Spring application preloader
      [--skip-listen], [--no-skip-listen]                    # Don't generate configuration that depends on the listen gem
  -J, [--skip-javascript], [--no-skip-javascript]            # Skip JavaScript files
      [--skip-turbolinks], [--no-skip-turbolinks]            # Skip turbolinks gem
      [--skip-jbuilder], [--no-skip-jbuilder]                # Skip jbuilder gem
  -T, [--skip-test], [--no-skip-test]                        # Skip test files
      [--skip-system-test], [--no-skip-system-test]          # Skip system test files
      [--skip-bootsnap], [--no-skip-bootsnap]                # Skip bootsnap gem
      [--dev], [--no-dev]                                    # Set up the application with Gemfile pointing to your Rails checkout
      [--edge], [--no-edge]                                  # Set up the application with Gemfile pointing to Rails repository
      [--master], [--no-master]                              # Set up the application with Gemfile pointing to Rails repository master branch
      [--rc=RC]                                              # Path to file containing extra configuration options for rails command
      [--no-rc], [--no-no-rc]                                # Skip loading of extra configuration options from .railsrc file
      [--api], [--no-api]                                    # Preconfigure smaller stack for API only apps
      [--minimal], [--no-minimal]                            # Preconfigure a minimal rails app
  -B, [--skip-bundle], [--no-skip-bundle]                    # Don't run bundle install
  --webpacker, [--webpack=WEBPACK]                           # Preconfigure Webpack with a particular framework (options: react, vue, angular, elm, stimulus)
      [--skip-webpack-install], [--no-skip-webpack-install]  # Don't run Webpack install

Runtime options:
  -f, [--force]                    # Overwrite files that already exist
  -p, [--pretend], [--no-pretend]  # Run but do not make any changes
  -q, [--quiet], [--no-quiet]      # Suppress status output
  -s, [--skip], [--no-skip]        # Skip files that already exist

Rails options:
  -h, [--help], [--no-help]        # Show this help message and quit
  -v, [--version], [--no-version]  # Show Rails version number and quit

Description:
    The 'rails new' command creates a new Rails application with a default
    directory structure and configuration at the path you specify.

    You can specify extra command-line arguments to be used every time
    'rails new' runs in the .railsrc configuration file in your home directory,
    or in $XDG_CONFIG_HOME/rails/railsrc if XDG_CONFIG_HOME is set.

    Note that the arguments specified in the .railsrc file don't affect the
    defaults values shown above in this help message.

Example:
    rails new ~/Code/Ruby/weblog

    This generates a skeletal Rails installation in ~/Code/Ruby/weblog.
</pre>
</details>

### System configuration
**Rails version**: 6.1.0

**Ruby version**: 2.7.2

It seems like there's a bad assumption in `Rails::Command#invoke` where we think the `command_name` could be `--help` but it can't be at the root (`rails --help` without a subcommand) since this conditional will always assign `command_name` to `"application"`: 
https://github.com/rails/rails/blob/95a8c83d2ec7aa34ab127f7fa6def4d8891eb662/railties/lib/rails/command.rb#L31-L41

Since the `command_name` is set (incorrectly IMO) to `"application"`, a call to `rails --help` results in the instantiation of a `Rails::Command::ApplicationCommand`
https://github.com/rails/rails/blob/95a8c83d2ec7aa34ab127f7fa6def4d8891eb662/railties/lib/rails/command.rb#L48-L51 

When `#perform` is called on that instance, the `help` method "punts" to `Rails::Generators::AppGenerator` with no arguments, which is why we end up with `rails new` usage instructions, which are somewhat helpful, although quite verbose and unrelated to our request (IMO, again). 

Try as I might I couldn't find how to print the `USAGE` from `HelpCommand` [which clearly exists](https://github.com/rails/rails/blob/95a8c83d2ec7aa34ab127f7fa6def4d8891eb662/railties/lib/rails/commands/help/USAGE#L1-L16) and which I definitely remember seeing in the past. Not with `rails` and no arguments (prints `rails new` usage), not with `rails help` no with `rails commands`. 

This is problematic because it means for someone new to the Rails command line, there's no way to list available commands. The only solution is to look up [the Rails Guide](https://guides.rubyonrails.org/command_line.html#command-line-basics) or trudge through the source (which seems unlikely for newcomers).

**Comment by kaspth** [comment]
Yes, let's fix this!

To be clear, is this from within an app directory or outside like the home directory?

**Comment by olivierlacan** [comment]
This example was from the rails command as a standalone RubyGem in my user home directory, not within a created Rails app root directory.

**Comment by aar0nr** [comment]
One argument to your proposal could be that none of the commands actually work, or are relevant outside of a Rails application. I'm not sure it's useful showing a new user a long list of commands that don't do anything, or that would result in errors if run. The only one that does work outside a Rails app is the `new` command, which is why you get this help. Once inside a Rails application, the `rails --help` will display the full list of commands.

**tldr;** I think it's working as intended.

**Comment by p8** [comment]
I can see it can be confusing to have the same command behave differently. 
Maybe we should have a ‘rails-new’ command instead?

**Comment by p8** [comment]
I’ve also had this happen inside an app directory when I deleted the bin directory, trying to debug binstubs. It would have been nice to have a ‘rails doctor’ command.

**Comment by olivierlacan** [comment]
@aar0nr I take your point but regardless of usability, it should be possible to ask Rails to list its available commands rather than ignore input.

If the `rails` command is ignoring `--help` for any reason it should say so explicitely: 

Example:
```
cd ~
rails --help
$ rails --help
Only the `rails new` command works outside of a rails application directory.

You can use `rails new --help` for details on how to create new rails app.

The most common rails commands are:
 generate     Generate new code (short-cut alias: "g")
 console      Start the Rails console (short-cut alias: "c")
 server       Start the Rails server (short-cut alias: "s")
 test         Run tests except system tests (short-cut alias: "t")
 test:system  Run system tests
 dbconsole    Start a console for the database specified in config/database.yml
              (short-cut alias: "db")

 new          Create a new Rails application. "rails new my_app" creates a
              new application called MyApp in "./my_app"

# ...
```

This is also a neat possible future tie in for any `rails new --interactive` options as mentioned in https://github.com/rails/rails/pull/39444 and https://discuss.rubyonrails.org/t/interactive-rails-new/74355/22.

**Comment by rafaelfranca** [comment]
@p8 do you want to work on that?

**Comment by olivierlacan** [comment]
@p8 Down to pair on it if you’d like too by the way. Been diving in that code a bit.

**Comment by p8** [comment]
@rafaelfranca sure!
@olivierlacan that’d be nice as well.
I’ve worked on a related PR https://github.com/rails/rails/pull/39561

**Comment by p8** [comment]
@olivierlacan I'm currently off for the holidays, but I'll have time to work on this in january.
